core/consensus/babe/impl/babe_impl.cpp

```diff
@@ -56,27 +56,57 @@ namespace kagome::consensus {
     BOOST_ASSERT(hasher_);
     BOOST_ASSERT(log_);
 
-    NextEpochDescriptor epoch_0_and_1_digest;
-    epoch_0_and_1_digest.randomness = genesis_configuration_->randomness;
-    epoch_0_and_1_digest.authorities =
-        genesis_configuration_->genesis_authorities;
-    epoch_storage_->addEpochDescriptor(0, epoch_0_and_1_digest);
-    epoch_storage_->addEpochDescriptor(1, epoch_0_and_1_digest);
+    NextEpochDescriptor init_epoch_desc;
+    init_epoch_desc.randomness = genesis_configuration_->randomness;
+    init_epoch_desc.authorities = genesis_configuration_->genesis_authorities;
+    [[maybe_unused]] bool init_epoch_desc_ok =
+        !!epoch_storage_->addEpochDescriptor(0, init_epoch_desc)
+        && !!epoch_storage_->addEpochDescriptor(1, init_epoch_desc);
+
+    BOOST_ASSERT(init_epoch_desc_ok);
   }
 
-  void BabeImpl::runGenesisEpoch() {
-    Epoch genesis_epoch;
-    genesis_epoch.epoch_index = 0;
-    genesis_epoch.authorities = genesis_configuration_->genesis_authorities;
-    genesis_epoch.randomness = genesis_configuration_->randomness;
-    genesis_epoch.epoch_duration = genesis_configuration_->epoch_length;
-    genesis_epoch.start_slot = 0;
+  void BabeImpl::start() {
+    auto &&[best_block_number, best_block_hash] = block_tree_->deepestLeaf();
+    log_->debug("Babe run on block with number {} and hash {}",
+                best_block_number,
+                best_block_hash);
 
-    next_slot_finish_time_ =
-        clock_->now() + genesis_configuration_->slot_duration;
+    if (__builtin_expect(best_block_number == 0, false)) {
+      auto epoch_digest_res = epoch_storage_->getEpochDescriptor(0);
+      if (not epoch_digest_res) {
+        log_->error("Last epoch digest does not exist for initial epoch");
+      }
+      auto &&epoch_digest = epoch_digest_res.value();
+
+      auto slot_duration = genesis_configuration_->slot_duration;
 
-    current_state_ = BabeState::SYNCHRONIZED;
-    runEpoch(genesis_epoch, next_slot_finish_time_);
+      Epoch epoch;
+      epoch.epoch_index = 0;
+      epoch.authorities = epoch_digest.authorities;
+      epoch.randomness = epoch_digest.randomness;
+      epoch.epoch_duration = genesis_configuration_->epoch_length;
+      // clang-format off
+	    // TODO(xDimon): remove one of next two variants by decision about way of slot enumeration
+	    //  variant #1
+//      epoch.start_slot = clock_->now().time_since_epoch() / slot_duration;
+//      auto starting_slot_finish_time = std::chrono::system_clock::time_point{}
+//                                       + slot_duration * epoch.start_slot
+//                                       + slot_duration;
+
+	    // TODO(xDimon): remove one of next two variants by decision about way of slot enumeration
+	    //  variant #2
+      epoch.start_slot = 0;
+      auto starting_slot_finish_time = std::chrono::system_clock::now()
+                                       + slot_duration * epoch.start_slot
```

Comment:
hm.. epoch.star_slot == 0. Why we do + 0?

Reply:
Sure. Will be changed